kustomization/kustomization_generator.go

```diff
@@ -0,0 +1,429 @@
+/*
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kustomization
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/kustomize/api/konfig"
+	"sigs.k8s.io/kustomize/api/krusty"
+	"sigs.k8s.io/kustomize/api/provider"
+	"sigs.k8s.io/kustomize/api/resmap"
+	kustypes "sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+	"sigs.k8s.io/yaml"
+
+	"github.com/fluxcd/pkg/apis/kustomize"
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	specField            = "spec"
+	targetNSField        = "targetNamespace"
+	patchesField         = "patches"
+	patchesSMField       = "patchesStrategicMerge"
+	patchesJson6902Field = "patchesJson6902"
+	imagesField          = "images"
+)
+
+type KustomizeGenerator struct {
+	kustomization unstructured.Unstructured
+}
+
+func NewGenerator(kustomization unstructured.Unstructured) *KustomizeGenerator {
+	return &KustomizeGenerator{
+		kustomization: kustomization,
+	}
+}
+
+// WriteFile generates a kustomization.yaml in the given directory if it does not exist.
+// It apply the flux kustomize resources to the kustomization.yaml and then write the
+// updated kustomization.yaml to the directory.
+// It returns the original kustomization.yaml.
+func (kg *KustomizeGenerator) WriteFile(dirPath string) ([]byte, error) {
```

Comment:
Are we going to overwrite people's kustomizations for a dry run, are we going to generate a kustomization.yaml if there is none? I guess we don't want any of this for `flux diff`. I think all this code can be used only in the controller.

Reply:
I actually use it in `flux diff`. We generate the `kustomization.yml` if it does not exist. And the return is `([]byte, error)`. This is use to pass the original content of the `kustomization.yaml`, so it can be restored if needed after all the transformation. Controller can just discard it, but the cli make sure to leave a clean state.